### ISSUE TITLE
No longer able to set header values to non-str

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6205]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#6205|
|Type|release blocker: regression|
|Created|2012-12-06 20:12:00Z|
|Branch|https://github.com/twisted/twisted/tree/non-bytes-headers-6205|

We previously allowed setting header values to non-str objects, like ints. Since the Python 3 port of http.py, that's no longer possible.

<details><summary>Searchable metadata</summary>

```
trac-id__6205 6205
type__release_blocker__regression release blocker: regression
reporter__radix radix
priority__normal normal
milestone__Twisted_12_3 Twisted-12.3
branch__branches_non_bytes_headers_6205 branches/non-bytes-headers-6205
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1354824720000000 1354824720000000
changetime__1355086145000000 1355086145000000
version__None None
owner__therve therve
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Adding to release milestone as this is a regression.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

It would be helpful if you mentioned what API the change in behavior relates to.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [36535]) Branching to 'non-bytes-headers-6205'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

The change was in http.Request.write. We used to force cast to string, now we just pass the data to writeSequence directly. The branch tries to change it back, but it's a bit gross. Maybe we should deprecate it too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.

  1. `setRawHeaders` never supported non-bytes arguments for the `name` argument (it required something with a `lower` method and if it were unicode, interpolating it in `write` later would produce unicode which would break the transport) and I don't think it should start now.
  1. The `.encode('ascii')` call in `write` is just what `networkString` is for.
  1. Also the comment in `write` is indented weirdly.
  1. I agree this should be deprecated.  It was never intended as a feature.

Please merge after these are fixed and buildbot says it works everywhere.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix commented|
|-|-|

I will let you judge for yourself, but I think there is a very uncommon incompatibility remaining.

Previously (before the py3k porting work), an object that had a __str__ that returned non-ascii bytes would work (meaning those bytes would be sent directly over the wire as returned), and now, in your branch, it won't. You'll get an ascii decoding or encoding error, since you're converting to unicode and then back to bytes.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/non-bytes-headers-6205 with the build.

I couldn't use networkString as far I can tell. Maybe I'm missing something, so I put it back for a round of review. Everything else handled.

Radix: maybe you're right. Considering it was never explicitly supported and it's more work, I'd wait for someone to complain.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

  1. You can use `networkString`.  You may have just had the impression that it was going to be more useful than it really is.  Instead of:

```python
value = ('%s' % (value,)).encode('ascii')
```

I suggest:

```python
value = networkString('%s' % (value,))
```

This just has the benefit of being more self-documented about why the value is being encoded to ascii - because it is a string that belongs on the network.

Something else I just noticed is that this doesn't support `unicode` values though (even ascii-only unicode).  Of course, the old version didn't support unicode either, so that's just fine.  It's kinda funny, since if I didn't know that this bug was filed because someone was passing an integer, I would have guessed the most likely incorrect type to be passed would be unicode.

  2. `test_nonByteHeaderValue` depends on dictionary iteration order, I think.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [36539]) Merge non-bytes-headers-6205

Author: therve
Reviewer: exarkun
Fixes: [#6205](https://github.com/twisted/twisted/issues/6205)

Re-allow usage of non-bytes header values in Request.write, as it was before
the py3 refactoring, but deprecate it.
